### PR TITLE
RFC: DFSANE algorithm

### PR DESCRIPTION
**Pull request by sglyon**
NOTE: This builds on top of #95 

In this PR I have taken a first stab at  implementing a derivative free method known as DFSANE. My implementation is largely based off the version in scipy, with some consultation of the papers below.

There are a few things that need to be addressed here before merging:

- Change to handling of keyword arguments in `nlsolve` methods. In this PR I chose to just splat them in all but one of the methods, and hand them off to the "main" method that calls the underlying solver. In this "main" method the common kwargs for all methods are type checked. The `df_sane` method allows for a few more parameters to be tuned, so right now they just get splatted even in this main method and handed on to a  `df_sane` function that does type checking.
- Addition of DataStructures as a dependency
- Inclusion of two non-monotone line search algorithms in the file df_sane.jl. Do these belong here, or are the potentially more generally applicable and candidates for moving to LineSearches.jl? cc @anriseth @pkofod 
- What do do about the scipy license? NLsolve.jl is MIT licensed. Scipy is BSD. I hope I'm ok in just making an attribution at the top of the file and copy/paste the scipy license, but I'm not sure on this one.
- More testing. 
    - I'm not directly testing the line search methods
    - I've just included df_sane as another algorithm choice in the minpack tests. It would be better to see where else it can succeed and try there.

### References

La Cruz, W. (2012). A projected derivative-free algorithm for nonlinear equations with convex constraints. Optimization Methods and Software, 29(1), 24–41. http://doi.org/10.1080/10556788.2012.721129

Cheng, W., & Li, D. H. (2009). A derivative-free nonmonotone line search and its application to the spectral residual method. IMA Journal of Numerical Analysis, 29(3), 814–825. http://doi.org/10.1093/imanum/drn019

Cruz, W. L., & Raydan, M. (2003). Nonmonotone Spectral Methods for Large-Scale Nonlinear Systems. Optimization Methods and Software, 18(5), 583–599. http://doi.org/10.1080/10556780310001610493

The file [scipy/optimize/_spectral.py](https://github.com/scipy/scipy/blob/2b7ce3919688d37f2620425ab6f87ad172f7425f/scipy/optimize/_spectral.py)

### CODE REVIEW

**Comment by pkofod** [comment]
We're going to need a proper (but simple ofc) description of what is in this pkg soon imo. Could you add a short intro to this method somewhere? Not like a derivation or convergence theory, just so a user or dev has an idea about what is going on. Could be a md file in a temp doc folder or something like that until we get docs up and running.

**Comment by pkofod** [comment]
> Inclusion of two non-monotone line search algorithms in the file df_sane.jl. To these belong here, or are the potentially more generally applicable and candidates for moving to LineSearches.jl?

I havn't looked at the code yet, but unless there's something special about the code that really ties it to this method, then I say: by all means put it in `LineSearches.jl`!

**Comment by pkofod** [comment]
We need a licensing expert to weigh in on that part, as I have no idea about these matters.

**Comment by anriseth** [comment]
> Inclusion of two non-monotone line search algorithms in the file df_sane.jl. Do these belong here, or are the potentially more generally applicable and candidates for moving to LineSearches.jl?

It could probably go into LineSearches. Great to have more choices for people, as in my experience there is never one particular line search that is the *best*. Your implementations of linesearches as types are also a good push/inspiration for how we may implement https://github.com/JuliaNLSolvers/LineSearches.jl/issues/9 :smiley: 

Out of interest: Do they assume that the objective is of the form `f(x) = \|G(x)\|_2^2` for some vector field `G(x)`?

**Comment by sglyon** [comment]
That's a really good question. I know that is what is used in this implementation, but I don't know if that is necessary. The last two papers I referred to should contain that information

**Comment by pkofod** [comment]
> Out of interest: Do they assume that the objective is of the form f(x) = \|G(x)\|_2^2 for some vector field G(x)?

Yes, so this should probably be well-documented if we put them in LineSearches

**Comment by pkofod** [comment]
What is DataStructures used for btw? I seem to be missing it.

**Comment by sglyon** [comment]
I will definitely have to document things when I find the time. I opened this to see if anyone had any early comments, but it is not in good enough shape to consider merging.

I'm not exactly sure when I will get around to tidying this up. 

Looks like you've figured out why I used DataStructures

File: src/df_sane.jl [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+                    σ_eps::Float64=1e-10,
+                    maxfev::Int=1_000,
+                    ftol::Float64=convert(T, 1e-10),
+                    xtol::Float64=1e-15,
+                    iterations::Int=3_000,
+                    store_trace::Bool=false,
+                    show_trace::Bool=false,
+                    extended_trace::Bool=false)
+
+    lsT = linesearch == :cruz ? CruzLineSearch : ChengLineSearch
```
Comment (pkofod) [inline src/df_sane.jl]: Why the symbols when you have a type to work off of?

Comment (sglyon) [inline src/df_sane.jl]: Good call, that should be changed.

**Comment by anriseth** [comment]
LineSearches are now using the callable-types approach, so it should be quite straight-forward to merge your new linesearches (if we agreed that they belong there)

**Comment by sglyon** [comment]
That's great, thanks for the heads up!

**Comment by cortner** [comment]
Would love to have non-monotone linesearch in LineSearches!

**Comment by pkofod** [comment]
@sglyon is it realistic that you'll come back and finish this at some point?

**Comment by sglyon** [comment]
Haha yep, it is realistic -- though I can't promise exactly when.

I would appreciate some help moving the line searches over to LineSearches.jl. I would be able to clean up the NLsolve side of the algorithm after that. 

Could anyone help with the movement of the linesearch algorithms?

**Comment by anriseth** [comment]
> Could anyone help with the movement of the linesearch algorithms?

We are looking at changing the LineSearches interface to do univariate optimization on an interval. It will hopefully be easier to move the linesearch algorithms after that is done.

Ref:
- https://github.com/JuliaNLSolvers/LineSearches.jl/issues/79#issuecomment-365389912 
- https://github.com/JuliaNLSolvers/Optim.jl/pull/543

**Comment by pkofod** [comment]
Yes I will have a pr to be discussed soon

**Comment by pkofod** [comment]
https://github.com/pkofod/NLSolvers.jl/blob/65034bb76cbf91e22784bd0156f28b85416adb2f/src/nlsolve/spectral/dfsane.jl has it if anyone needs it

**Pull request closed without merging**